# Import Packages and load in data

In [1]:
## Import packages

import xarray as xr

## Load in data

ds_cGAN = xr.open_dataset('../datasets/east_africa-cgan_forecast-20240422.nc')


## What is in the data?

#### 1) What dimensions, coordinates and variables are there?
#### 2) What are the dimensions for the variables & in what order do they appear?
#### 3) Let's try select a member and some lon-lat points


In [2]:
## What does the data look like?

ds_cGAN

<xarray.Dataset> Size: 108MB
Dimensions:          (latitude: 384, longitude: 352, member: 50, time: 1,
                      valid_time: 4)
Coordinates:
  * latitude         (latitude) float32 2kB -13.65 -13.55 -13.45 ... 24.55 24.65
  * longitude        (longitude) float32 1kB 19.15 19.25 19.35 ... 54.15 54.25
  * member           (member) int32 200B 1 2 3 4 5 6 7 ... 44 45 46 47 48 49 50
  * time             (time) datetime64[ns] 8B 2024-04-22T00:00:57.096667136
Dimensions without coordinates: valid_time
Data variables:
    fcst_valid_time  (time, valid_time) datetime64[ns] 32B ...
    precipitation    (time, member, valid_time, latitude, longitude) float32 108MB ...
Attributes:
    description:  GAN 6-hour rainfall ensemble members in the ICPAC region.

In [14]:
## Check which members are available to select

ds_cGAN.member.values

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50],
      dtype=int32)

In [16]:
## Select a given member or set of members using sel (i.e., by index label)

ds_cGAN.sel({'member':[24]})

<xarray.Dataset> Size: 2MB
Dimensions:          (latitude: 384, longitude: 352, member: 1, time: 1,
                      valid_time: 4)
Coordinates:
  * latitude         (latitude) float32 2kB -13.65 -13.55 -13.45 ... 24.55 24.65
  * longitude        (longitude) float32 1kB 19.15 19.25 19.35 ... 54.15 54.25
  * member           (member) int32 4B 24
  * time             (time) datetime64[ns] 8B 2024-04-22T00:00:57.096667136
Dimensions without coordinates: valid_time
Data variables:
    fcst_valid_time  (time, valid_time) datetime64[ns] 32B 2024-04-23T06:01:2...
    precipitation    (time, member, valid_time, latitude, longitude) float32 2MB ...
Attributes:
    description:  GAN 6-hour rainfall ensemble members in the ICPAC region.

In [17]:
## Select the first lon and lat point using isel (i.e. by index integer)

ds_cGAN.isel({'longitude':[0], 'latitude':[0]})


The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


<xarray.Dataset> Size: 1kB
Dimensions:          (latitude: 1, longitude: 1, member: 50, time: 1,
                      valid_time: 4)
Coordinates:
  * latitude         (latitude) float32 4B -13.65
  * longitude        (longitude) float32 4B 19.15
  * member           (member) int32 200B 1 2 3 4 5 6 7 ... 44 45 46 47 48 49 50
  * time             (time) datetime64[ns] 8B 2024-04-22T00:00:57.096667136
Dimensions without coordinates: valid_time
Data variables:
    fcst_valid_time  (time, valid_time) datetime64[ns] 32B 2024-04-23T06:01:2...
    precipitation    (time, member, valid_time, latitude, longitude) float32 800B ...
Attributes:
    description:  GAN 6-hour rainfall ensemble members in the ICPAC region.

#### Note this would be the same as doing ds_cGAN.sel({'longitude':[19.15], 'latitude':[-13.65]})!


## Sorting the ensemble members

In [4]:
ds_cGAN.sortby('member')

<xarray.Dataset> Size: 108MB
Dimensions:          (latitude: 384, longitude: 352, member: 50, time: 1,
                      valid_time: 4)
Coordinates:
  * latitude         (latitude) float32 2kB -13.65 -13.55 -13.45 ... 24.55 24.65
  * longitude        (longitude) float32 1kB 19.15 19.25 19.35 ... 54.15 54.25
  * member           (member) int32 200B 1 2 3 4 5 6 7 ... 44 45 46 47 48 49 50
  * time             (time) datetime64[ns] 8B 2024-04-22T00:00:57.096667136
Dimensions without coordinates: valid_time
Data variables:
    fcst_valid_time  (time, valid_time) datetime64[ns] 32B ...
    precipitation    (time, member, valid_time, latitude, longitude) float32 108MB ...
Attributes:
    description:  GAN 6-hour rainfall ensemble members in the ICPAC region.

## Calculating mean, min and max

#### 1) We show for mean
#### 2) Now you try what would you do for min and max?

In [5]:
ds_cGAN.mean('member')

<xarray.Dataset> Size: 2MB
Dimensions:          (latitude: 384, longitude: 352, time: 1, valid_time: 4)
Coordinates:
  * latitude         (latitude) float32 2kB -13.65 -13.55 -13.45 ... 24.55 24.65
  * longitude        (longitude) float32 1kB 19.15 19.25 19.35 ... 54.15 54.25
  * time             (time) datetime64[ns] 8B 2024-04-22T00:00:57.096667136
Dimensions without coordinates: valid_time
Data variables:
    fcst_valid_time  (time, valid_time) datetime64[ns] 32B 2024-04-23T06:01:2...
    precipitation    (time, valid_time, latitude, longitude) float32 2MB 0.00...

## Calculating accumulated over the day

In [6]:
## Note that we have 4 valid times i.e., 6-hourly data

ds_cGAN.fcst_valid_time

<xarray.DataArray 'fcst_valid_time' (time: 1, valid_time: 4)> Size: 32B
array([['2024-04-23T06:01:24.114096128', '2024-04-23T11:58:44.590837760',
        '2024-04-23T18:00:39.945486336', '2024-04-23T23:58:00.422227968']],
      dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 8B 2024-04-22T00:00:57.096667136
Dimensions without coordinates: valid_time

In [7]:
ds_cGAN.sum('valid_time')

<xarray.Dataset> Size: 27MB
Dimensions:        (latitude: 384, longitude: 352, member: 50, time: 1)
Coordinates:
  * latitude       (latitude) float32 2kB -13.65 -13.55 -13.45 ... 24.55 24.65
  * longitude      (longitude) float32 1kB 19.15 19.25 19.35 ... 54.15 54.25
  * member         (member) int32 200B 1 2 3 4 5 6 7 8 ... 44 45 46 47 48 49 50
  * time           (time) datetime64[ns] 8B 2024-04-22T00:00:57.096667136
Data variables:
    precipitation  (time, member, latitude, longitude) float32 27MB 0.02385 ....